src/hooks/useDimensions/index.ts

```diff
@@ -0,0 +1,37 @@
+import { MutableRefObject, useLayoutEffect, useMemo, useState } from "react";
+
+interface DimensionState {
+  width: number;
+  height: number;
+}
+
+/**
+ * useDimensions returns the pixel width and height of a given ref.
+ * The ref should be applied to an HTML Element.
+ * @param ref The ref of the element to track the dimensions of.
+ * @returns An object containing the current height and width of the ref.
+ */
+export const useDimensions = (ref: MutableRefObject<HTMLElement>) => {
+  const [state, setState] = useState<DimensionState>({ width: 0, height: 0 });
+
+  const observer = useMemo(
+    () =>
+      new ResizeObserver((entries) => {
+        requestAnimationFrame(() => {
+          const { width, height } = entries[0]?.contentRect ?? {};
+          setState({ width, height });
```

Comment:
I'm not entirely sure if this is preferable to a debounce. Especially considering setState calls are async so there is no guarantee that they will actually execute within the requestAnimationFrame.  We can monitor bugsnag if this comes up.